### ISSUE TITLE
Docstring default values for mutable arguments.

### ISSUE THREAD

**Issue by vinRosso** [mkdocstrings/mkdocstrings#577]
### Discussed in https://github.com/mkdocstrings/mkdocstrings/discussions/527

<div type='discussions-op-text'>

<sup>Originally posted by **ZhiyuanChen** January 29, 2023</sup>
Set default in arguments to mutable objects could cause [horrible errors](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/).
A general solution is to set default to None, and then assign correct default value in the function body. 
Although the correct default value should be properly addressed in docs to avoid ambiguation. 

However, it seems mkdocstring is not recognising the default values specified, like this:
<img width="849" alt="image" src="https://user-images.githubusercontent.com/28757366/215310152-217bd1e5-4b1f-4a23-bafb-6f722ac9c922.png">
</div>

### Possible solution
Adding a *Defaults* section to the docstring could be the best solution so far.
```py
def google_docstring(a, b=False, c=None):
    """ Description.
    
    Args:
        a (int): The first parameter.
        b (bool): The second parameter.
        c (list | None): The third parameter.

    Defaults:
        c: [1, 2, 3, 4]
    """

def numpy_docstring(a, b=False, c=None):
    """ Description.
    
    Parameters
    ----------
    a : int
        The first parameter.
    b : bool
        The second parameter.
    c : list | None
        The third parameter.

    Defaults
    --------
    c : [1, 2, 3, 4]
    """

def sphinx_docstring(a, b=False, c=None):
    """ Description.
    
    :param a: The first parameter.
    :type a: int

    :param b: The second parameter.
    :type b: bool

    :param c: The second parameter.
    :type c: list | None
    :default c: [1, 2, 3, 4]
    """
```

**Comment by pawamoy** [comment]
I've thought about it a bit more, and in fact I don't like having a new "Defaults" section. It would act as some kind of metadata section, that we would have to merge into other ones (Parameters, Other Parameters) and not render. So it would behave differently than every other section.

The only way I can see to parse a default value correctly within the parameter line is:

```
c (list | None, default=[1, 2, 3, 4]): The third parameter.
```

...and to `compile` that as a function call, to an AST, to fetch the default value of the `default` parameter, which is our default value of `c`. That means: no textual description for default values, only valid Python syntax.

We could do the same thing for Numpydoc, which already supports `default=0` (but also `default 0` and `default: 0` which would have to be dropped).

**Comment by vinRosso** [comment]
I can see that it would be odd, since Griffe is already branching off from official docstrings documentations... But eventually "Defaults" would be an extra metadata section, such as the "See Also" and "Methods" sections that you mentioned in your [numpydoc issue](https://github.com/numpy/numpydoc/issues/463#issuecomment-1610106555), right?.

Right now I'm also using arbitrary text to document some types in Google, as it's easier to read. So my worry is that compiling the whole thing as Python code would loose this advantage. I'm not a professional programmer, so there may be some better way to define a list of strings for example, but I hope that you get my point.
```py
#e.g.
c (str[]): description`
c (list[str]): description
```
So I still think that adding the "Defaults" custom section would be the best way, as it's docstring type agnostic and allows for more freedom... But in the end, for me at least, it doesn't really matter what solution you're going to take, since anyone of them would be a nice extra feature to have, and I can adapt to whatever you're going to choose.

**Comment by pawamoy** [comment]
> such as the "See Also" and "Methods" sections that you mentioned in your https://github.com/numpy/numpydoc/issues/463#issuecomment-1610106555, right?.

No, See Also and Methods sections can be rendered :slightly_smiling_face:  They don't act like metadata.

> Right now I'm also using arbitrary text to document some types in Google

Oh OK, it's true that it's not explicitly forbidden. Thanks for letting me know.

> `c (list[str]): description`

Definitely the right way to annotate as a list of string :smile: 

> But in the end, for me at least, it doesn't really matter what solution you're going to take, since anyone of them would be a nice extra feature to have, and I can adapt to whatever you're going to choose.

Noted, thanks!
